cmd/vlog_health_checker/main.go

```diff
@@ -0,0 +1,158 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package main
+
+import (
+	"encoding/binary"
+	"flag"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/log"
+	"github.com/pingcap/tidb-binlog/pump/storage"
+	pb "github.com/pingcap/tipb/go-binlog"
+	"github.com/syndtr/goleveldb/leveldb"
+	"go.uber.org/zap"
+)
+
+var tsKeyPrefix = []byte("ts:")
+
+func main() {
+	// Parse args
+	kvDir := flag.String("kv_dir", "", "path of the pump kv directory")
+	repair := flag.Bool("repair", false, "repair the offsets in LevelDB")
+	flag.Parse()
+	path := flag.Arg(0)
+
+	// Open log file
+	log.Info("Target vlog", zap.String("path", path))
+	fid, err := storage.ParseFid(path)
+	if err != nil {
+		log.Fatal("Failed to parse fid", zap.Error(err), zap.String("path", path))
+	}
+	log.Info("Parse fid", zap.Uint32("fid", fid))
+	logFile, err := storage.NewLogFile(fid, path)
+	if err != nil {
+		log.Fatal("Failed to open log", zap.Error(err), zap.String("path", path))
+	}
+	log.Info("Log opened")
+
+	// Open LevelDB
+	metadb, err := storage.OpenMetadataDB(*kvDir, nil)
+	if err != nil {
+		log.Fatal("Failed to open LevelDB", zap.String("path", *kvDir), zap.Error(err))
+	}
+	defer func() {
+		err := metadb.Close()
+		if err != nil {
+			log.Error("Failed to close metadb", zap.Error(err))
+		}
+	}()
+	log.Info("LevelDB opened")
+
+	var corruptionPos int64 = -1
+	checked := make(map[int64]struct{}, 64)
+	logFile.Scan(
+		0,
+		func(vp storage.ValuePointer, record *storage.Record) error {
+			bl, err := record.GetBinlog()
+			if err != nil {
+				log.Fatal("Failed to get binlog", zap.Int64("offset", vp.Offset), zap.Error(err))
+			}
+			// Ignore rollback binlogs
+			if bl.Tp == pb.BinlogType_Rollback {
+				return nil
+			}
+			tso := getTSO(bl)
+			if _, ok := checked[tso]; ok {
+				log.Info("Skip duplicate binlog", zap.Int64("tso", tso))
+				return nil
+			}
+			checked[tso] = struct{}{}
+			key := encodeKey(bl)
+
+			if *repair {
+				data, err := vp.MarshalBinary()
+				if err != nil {
+					log.Fatal("Failed to marshal pointer", zap.Error(err), zap.Int64("offset", vp.Offset))
+				}
+				if err := metadb.Put(key, data, nil); err != nil {
+					log.Fatal("Failed to repair pointer", zap.Error(err), zap.Int64("offset", vp.Offset))
+				}
+				log.Info("Fixed", zap.Int64("offset", vp.Offset), zap.Binary("key", key))
+				return nil
+			}
+
+			pointer, err := getSavedPointer(metadb, key)
+			if err != nil {
+				log.Error("Failed to get saved data", zap.Error(err), zap.Binary("key", key))
+				return nil
+			}
+
+			if pointer.Fid != fid {
+				// Reading a different Fid, this should never happen
+				log.Fatal("Fid mismatch", zap.Uint32("get", pointer.Fid), zap.Uint32("expected", fid))
+			}
+
+			if pointer.Offset != vp.Offset {
```

Comment:
s/if/else if/

Reply:
The `if` block above called `log.Fatal` which called `os.Exit(1)` for us.